### PR TITLE
SNOW-922795 Added third-party-licenses to .zip distribution

### PR DESCRIPTION
**Pull request by sfc-gh-mbobowski**
This PR adds third-party-license files into Confluent zip distribution.

The files are copied directly from jars (if possible) or inserted into ADDITIONAL_LICENSES from the script (see ADDITIONAL_LICENSES_MAP). If dependency does not provide license file in jar and is not present in ADDITIONAL_LICENSES_MAP then the build fails.

On the sreenshot below you can see what is the difference between old and new version of zip file. It needs to be approve by legal before merging.
Screenshot 2024-02-06 at 16 01 22

### CODE REVIEW

**Review by sfc-gh-xhuang — APPROVED** [review]

**Comment by sfc-gh-mfei** [comment]
What happens if a license is declared (e.g., SPDX identifier) but no license file is found?

**Comment by sfc-gh-mbobowski** [comment]
@sfc-gh-mfei What do you mean by `license is declared`? Declared in the [process_licenses.py script](https://github.com/snowflakedb/snowflake-kafka-connector/pull/784/files#diff-323e67e759aba464761b3b930050725c9445da7b7ed3f4274573880e952b140eR26-R71)? Or declared in the dependency jar?

The script basically looks for the files named "META-INF/LICENSE.txt", "META-INF/LICENSE", "META-INF/LICENSE.md" and extracts them. If the license is declared in any other way then it cannot be extracted and put inside the zip file. This is consistent with `snowflake-ingest-java`.

Please, let me know if you have any questions.

**Review by sfc-gh-tzhang — APPROVED** [review]
lgtm, i would ask @sfc-gh-lsembera and @sfc-gh-mfei for a quick look, thanks!

**Comment by sfc-gh-mfei** [comment]
> @sfc-gh-mfei What do you mean by `license is declared`? Declared in the [process_licenses.py script](https://github.com/snowflakedb/snowflake-kafka-connector/pull/784/files#diff-323e67e759aba464761b3b930050725c9445da7b7ed3f4274573880e952b140eR26-R71)? Or declared in the dependency jar?
> 
> The script basically looks for the files named "META-INF/LICENSE.txt", "META-INF/LICENSE", "META-INF/LICENSE.md" and extracts them. If the license is declared in any other way then it cannot be extracted and put inside the zip file. This is consistent with `snowflake-ingest-java`.
> 
> Please, let me know if you have any questions.

Sometimes a dependency will not include an actual license file in meta-inf, but it will "declare" a license in the pom:

  <licenses>
    <license>
      <name>Apache-2.0</name>
      <url>https://www.apache.org/licenses/LICENSE-2.0.txt</url>
    </license>
  </licenses>
  
  In those cases (where a license is declared in the pom, but no actual license file is included), we should have a file where we dump the names of those dependencies and the license they identify.  Does that make sense?  Also happy to chat live if needed.

File: scripts/process_licenses.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+EDL_10_LICENSE = "EDL 1.0"
+MIT_LICENSE = "The MIT License"
+GO_LICENSE = "The Go license"
+BOUNCY_CASTLE_LICENSE = "Bouncy Castle License"
+
+# The SDK does not need to include licenses of dependencies, which aren't shaded
+IGNORED_DEPENDENCIES = {"net.snowflake:snowflake-jdbc", "org.slf4j:slf4j-api"}
+
+# List of dependencies, which don't ship with a license file.
+# Only add a new record here after verifying that the dependency JAR does not contain a license!
```
Comment (sfc-gh-japatel) [inline scripts/process_licenses.py:39]: might be worth calling it out in README?

File: scripts/process_licenses.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+                    missing_licenses_str += f"{dependency_lookup_key}: {license_name}\n"
+                else:
+                    raise Exception(f"The dependency {dependency_lookup_key} does not ship a license file, but neither is it not defined in ADDITIONAL_LICENSES_MAP")
+
+    with open(Path(target_dir, "ADDITIONAL_LICENCES"), "w") as additional_licenses_handle:
+        additional_licenses_handle.write(missing_licenses_str)
+
+    if dependency_count < 30:
+        raise Exception(f"Suspiciously low number of dependency JARs detected in {dependency_jars_path}: {dependency_count}")
+    print("License generation finished")
```
Comment (sfc-gh-japatel) [inline scripts/process_licenses.py:149]: I dont see this print statements in confluent runs in github actions, is that expected?

**Review by sfc-gh-japatel — APPROVED** [review]
lgtm, couple of non blocking comments.

**Comment by sfc-gh-mbobowski** [comment]
> > @sfc-gh-mfei What do you mean by `license is declared`? Declared in the [process_licenses.py script](https://github.com/snowflakedb/snowflake-kafka-connector/pull/784/files#diff-323e67e759aba464761b3b930050725c9445da7b7ed3f4274573880e952b140eR26-R71)? Or declared in the dependency jar?
> > The script basically looks for the files named "META-INF/LICENSE.txt", "META-INF/LICENSE", "META-INF/LICENSE.md" and extracts them. If the license is declared in any other way then it cannot be extracted and put inside the zip file. This is consistent with `snowflake-ingest-java`.
> > Please, let me know if you have any questions.
> 
> Sometimes a dependency will not include an actual license file in meta-inf, but it will "declare" a license in the pom (sorry for the apostrophes, but I don't know how to stop GitHub from parsing it):
> 
> > '<'licenses'>'
> > '<'license'>'
> > '<'name'>'Apache-2.0'<'/name'>'
> > '<'url'>'https://www.apache.org/licenses/LICENSE-2.0.txt'<'/url'>'
> > '<'/license'>'
> > '<'/licenses'>'
> 
> In those cases (where a license is declared in the pom, but no actual license file is included), we should have a file where we dump the names of those dependencies and the license they identify. Does that make sense? Also happy to chat live if needed.

@sfc-gh-mfei Thanks, now it is clear to me. If there is no license file in meta-inf the dependency will be posted in ADDITIONAL_LICENSES. The content of this file is:
```
org.bouncycastle:bcpkix-fips: Bouncy Castle Licence
org.bouncycastle:bc-fips: Bouncy Castle Licence
org.apache.kafka:kafka-clients: Apache License 2.0
net.snowflake:snowflake-ingest-sdk: Apache License 2.0
io.confluent:kafka-schema-registry-client: Apache License 2.0
io.swagger.core.v3:swagger-annotations: Apache License 2.0
io.confluent:kafka-avro-serializer: Apache License 2.0
io.confluent:kafka-schema-serializer: Apache License 2.0
io.confluent:logredactor: Apache License 2.0
com.google.re2j:re2j: The Go license
io.confluent:logredactor-metrics: Apache License 2.0
com.eclipsesource.minimal-json:minimal-json: The MIT License
io.confluent:kafka-connect-avro-converter: Apache License 2.0
io.confluent:kafka-connect-avro-data: Apache License 2.0
io.confluent:kafka-schema-converter: Apache License 2.0
io.confluent:common-utils: Apache License 2.0
io.dropwizard.metrics:metrics-core: Apache License 2.0
io.dropwizard.metrics:metrics-jmx: Apache License 2.0
com.google.guava:failureaccess: Apache License 2.0
com.google.guava:listenablefuture: Apache License 2.0
com.google.code.findbugs:jsr305: Apache License 2.0
com.google.errorprone:error_prone_annotations: Apache License 2.0
com.google.j2objc:j2objc-annotations: Apache License 2.0
dev.failsafe:failsafe: Apache License 2.0
```

**Review by sfc-gh-lsembera — APPROVED** [review]

**Review by sfc-gh-gjachimko — APPROVED** [review]

File: pom_confluent.xml [hunk 3]
```diff
@@ -49,6 +49,9 @@
         <skipTests>false</skipTests>
         <skipUnitTests>${skipTests}</skipUnitTests>
         <skipIntegrationTests>${skipTests}</skipIntegrationTests>
+        <license.processing.dependencyJarsDir>${project.build.directory}/dependency-jars</license.processing.dependencyJarsDir>
+        <license.processing.dependencyListFile>${project.build.directory}/dependency-list.txt</license.processing.dependencyListFile>
+        <license.processing.targetDir>${project.build.directory}/../licenses</license.processing.targetDir>
```
Comment (sfc-gh-achyzy) [inline pom_confluent.xml:54]: I prefer if you not pollute the main directory with generates sources. Target directory in maven is specifically for that. 
Also, if you run `clean` the `licenses` directory is not cleaned up.

**Review by sfc-gh-achyzy — APPROVED** [review]
LGTM. 
But I would really appreciate if you fix the cleanup for `licenses` directory.

File: pom_confluent.xml [hunk 4]
```diff
@@ -81,6 +84,76 @@
     </distributionManagement>
     <build>
         <plugins>
+            <!--
+            Copy all project dependencies to target/dependency-jars. License processing Python script will look here for
+            license files of SDK dependencies.
```
Comment (sfc-gh-dseweryn) [inline pom_confluent.xml:89]: Nit: remove ambiguity
```suggestion
            Copy all project dependencies to target/dependency-jars directory. License processing Python script will check
            only this directory for license files of SDK dependencies.
```

**Review by sfc-gh-dseweryn — APPROVED** [review]
LGTM, one comment on clarity

**Comment by sfc-gh-mfei** [comment]
> > > @sfc-gh-mfei What do you mean by `license is declared`? Declared in the [process_licenses.py script](https://github.com/snowflakedb/snowflake-kafka-connector/pull/784/files#diff-323e67e759aba464761b3b930050725c9445da7b7ed3f4274573880e952b140eR26-R71)? Or declared in the dependency jar?
> > > The script basically looks for the files named "META-INF/LICENSE.txt", "META-INF/LICENSE", "META-INF/LICENSE.md" and extracts them. If the license is declared in any other way then it cannot be extracted and put inside the zip file. This is consistent with `snowflake-ingest-java`.
> > > Please, let me know if you have any questions.
> > 
> > 
> > Sometimes a dependency will not include an actual license file in meta-inf, but it will "declare" a license in the pom (sorry for the apostrophes, but I don't know how to stop GitHub from parsing it):
> > > '<'licenses'>'
> > > '<'license'>'
> > > '<'name'>'Apache-2.0'<'/name'>'
> > > '<'url'>'https://www.apache.org/licenses/LICENSE-2.0.txt'<'/url'>'
> > > '<'/license'>'
> > > '<'/licenses'>'
> > 
> > 
> > In those cases (where a license is declared in the pom, but no actual license file is included), we should have a file where we dump the names of those dependencies and the license they identify. Does that make sense? Also happy to chat live if needed.
> 
> @sfc-gh-mfei Thanks, now it is clear to me. If there is no license file in meta-inf the dependency will be posted in ADDITIONAL_LICENSES. The content of this file is:
> 
> ```
> org.bouncycastle:bcpkix-fips: Bouncy Castle Licence
> org.bouncycastle:bc-fips: Bouncy Castle Licence
> org.apache.kafka:kafka-clients: Apache License 2.0
> net.snowflake:snowflake-ingest-sdk: Apache License 2.0
> io.confluent:kafka-schema-registry-client: Apache License 2.0
> io.swagger.core.v3:swagger-annotations: Apache License 2.0
> io.confluent:kafka-avro-serializer: Apache License 2.0
> io.confluent:kafka-schema-serializer: Apache License 2.0
> io.confluent:logredactor: Apache License 2.0
> com.google.re2j:re2j: The Go license
> io.confluent:logredactor-metrics: Apache License 2.0
> com.eclipsesource.minimal-json:minimal-json: The MIT License
> io.confluent:kafka-connect-avro-converter: Apache License 2.0
> io.confluent:kafka-connect-avro-data: Apache License 2.0
> io.confluent:kafka-schema-converter: Apache License 2.0
> io.confluent:common-utils: Apache License 2.0
> io.dropwizard.metrics:metrics-core: Apache License 2.0
> io.dropwizard.metrics:metrics-jmx: Apache License 2.0
> com.google.guava:failureaccess: Apache License 2.0
> com.google.guava:listenablefuture: Apache License 2.0
> com.google.code.findbugs:jsr305: Apache License 2.0
> com.google.errorprone:error_prone_annotations: Apache License 2.0
> com.google.j2objc:j2objc-annotations: Apache License 2.0
> dev.failsafe:failsafe: Apache License 2.0
> ```

Got it LGTM thanks!

File: pom_confluent.xml (hunk 3 above)
Comment (sfc-gh-mbobowski) [reply pom_confluent.xml:59]: Added cleanup for licenses directory

[the remark this answers is not in the document]
Comment (sfc-gh-mbobowski) [reply pom_confluent.xml:59]: No unfortunately there is no way to change this path. It is undocumented and the source code is unavailable. I wrote to Confluent to get it, but I do not want to hold the PR.

On the other hand I added cleanup and comment in pom.

File: pom_confluent.xml (hunk 4 above)
Comment (sfc-gh-mbobowski) [reply pom_confluent.xml:94]: Done

File: scripts/process_licenses.py (hunk 1 above)
Comment (sfc-gh-mbobowski) [reply scripts/process_licenses.py:39]: Added README section about third-parties

File: scripts/process_licenses.py (hunk 2 above)
Comment (sfc-gh-mbobowski) [reply scripts/process_licenses.py:149]: The logs are present but you have to go into the job logs
https://github.com/snowflakedb/snowflake-kafka-connector/actions/runs/7874696489/job/21484913973 and then open section `Build With Unit Test for Confluent`.

**Pull request merged**